File: backend/shopping/src/middlewares/auth.middleware.ts
import { Request, Response, NextFunction } from "express";

import { verifyAccessToken } from "../jwt";
import { fail } from "../result";

export function authenticate(req: Request, res: Response, next: NextFunction) {
  const authorization = req.headers.authorization;

  if(!authorization) {
    const result = fail("Invalid access token", 401);
    return res.status(result.status).json(result);
  };

  const parts = authorization.split(" ");

  if(parts.length !== 2 || parts[0] !== "Bearer" || !parts[1]) {
    const result = fail("Invalid access token", 401);
    return res.status(result.status).json(result);
  };

  const accessToken = parts[1];

  const payload = verifyAccessToken(accessToken);
  
  req.user = payload;
  next();
};

File: backend/auth/cmd/middlewares/auth.middleware.go
package middlewares

import (
  "Auth/cmd/config"
  "Auth/cmd/jwt"

  "errors"
  "strings"
  "github.com/gin-gonic/gin"
  "github.com/jackc/pgx/v5/pgconn"
);

func Authenticate(ctx *gin.Context) {
  authorization := ctx.GetHeader("Authorization");

  parts := strings.SplitN(authorization, " ", 2);

  if len(parts) != 2 || parts[0] != "Bearer" || parts[1] == "" {
    ctx.AbortWithStatusJSON(401, gin.H {
      "error": "Invalid access token",
	  });
	  return;
  };

  accessString := parts[1];

  payload, err := jwt.VerifyAccessToken(accessString, config.AccessSecret);

  if err != nil {
    ctx.AbortWithStatusJSON(401, gin.H {
      "error": "Invalid access token",
	  });
	  return;
  };

  ctx.Set("userId", payload.UserId);

  ctx.Next();
};

func UniqueHandler(ctx *gin.Context) {
  ctx.Next();

  for _, err := range ctx.Errors {
    var pgErr *pgconn.PgError;

    if errors.As(err.Err, &pgErr) && pgErr.Code == "23505" && pgErr.ConstraintName == "users_email_unique" {
      if !ctx.Writer.Written() {
        ctx.JSON(409, gin.H {
          "error": "An account with this email address already exists",
        });
      };
      return;
    };  
  };
};

func ErrorHandler(ctx *gin.Context) {
  ctx.Next();

  if len(ctx.Errors) > 0 {
    err := ctx.Errors.Last().Err

    if ctx.Writer.Written() {
      return;
    };

    ctx.JSON(400, gin.H {
      "error": err.Error(),
    });
    return;
  };

  if !ctx.Writer.Written() {
    ctx.JSON(500, gin.H {
      "error": "Internal server error",
    });
    return;
  };
};